plutus-core/plutus-core/src/PlutusCore/ParserCommon.hs

```diff
@@ -0,0 +1,202 @@
+      errorCode UnknownBuiltinType {}  = ErrorCode 5
+      errorCode BuiltinTypeNotAStar {} = ErrorCode 52
+      errorCode InvalidConstant {}     = ErrorCode 4
+
+type Error = Parsec.ParseError Char ParseError
+
+instance ShowErrorComponent ParseError where
+    showErrorComponent (UnknownBuiltinType ty) = "Unknown built-in type: " ++ T.unpack ty
+    showErrorComponent (BuiltinTypeNotAStar ty) =
+        "Expected a type of kind star (to later parse a constant), but got: " ++ T.unpack ty
+    showErrorComponent (InvalidConstant ty con) =
+        "Invalid constant: " ++ T.unpack con ++ " of type " ++ T.unpack ty
+
+topSourcePos :: SourcePos
+topSourcePos = initialPos "top"
+
+initial :: ParserState
+initial = ParserState M.empty
+
+-- | Return the unique identifier of a name.
+-- If it's not in the current parser state, map the name to a fresh id
+-- and add it to the state. Used in the Name parser.
+intern :: (MonadState ParserState m, PLC.MonadQuote m)
+    => T.Text -> m PLC.Unique
+intern n = do
+    st <- get
+    case M.lookup n (identifiers st) of
+        Just u -> return u
+        Nothing -> do
+            fresh <- PLC.freshUnique
+            let identifiers' = M.insert n fresh $ identifiers st
+            put $ ParserState identifiers'
+            return fresh
+
+type Parser = ParsecT ParseError T.Text (StateT ParserState PLC.Quote)
+instance (Stream s, PLC.MonadQuote m) => PLC.MonadQuote (ParsecT e s m)
+
+parse :: Parser a -> String -> T.Text -> Either (ParseErrorBundle T.Text ParseError) a
+parse p file str = PLC.runQuote $ parseQuoted p file str
+
+parseQuoted :: Parser a -> String -> T.Text -> PLC.Quote (Either (ParseErrorBundle T.Text ParseError) a)
+parseQuoted p file str = flip evalStateT initial $ runParserT p file str
+
+whitespace :: Parser ()
+whitespace = Lex.space space1 (Lex.skipLineComment "--") (Lex.skipBlockCommentNested "{-" "-}")
+
+lexeme :: Parser a -> Parser a
+lexeme = Lex.lexeme whitespace
+
+symbol :: T.Text -> Parser T.Text
+symbol = Lex.symbol whitespace
+
+lparen :: Parser T.Text
+lparen = symbol "("
+rparen :: Parser T.Text
+rparen = symbol ")"
+
+lbracket :: Parser T.Text
+lbracket = symbol "["
+rbracket :: Parser T.Text
+rbracket = symbol "]"
+
+lbrace :: Parser T.Text
+lbrace = symbol "{"
+rbrace :: Parser T.Text
+rbrace = symbol "}"
+
+inParens :: Parser a -> Parser a
+inParens = between lparen rparen
+
+inBrackets :: Parser a -> Parser a
+inBrackets = between lbracket rbracket
+
+inBraces :: Parser a -> Parser a
+inBraces = between lbrace rbrace
+
+isIdentifierChar :: Char -> Bool
+isIdentifierChar c = isAlphaNum c || c == '_' || c == '\''
+
+-- | Create a parser that matches the input word and returns its source position.
+-- This is for attaching source positions to parsed terms/programs.
+-- getSourcePos is not cheap, don't call it on matching of every token.
+reservedWord ::
+    -- | The word to match
+    T.Text -> Parser SourcePos
+reservedWord w = lexeme $ try $ do
```

Comment:
Good point! And thanks for the explanation!

Does `annWord` sound ok? It kinda returns the `ann` of things (`Term`, `Program`...).

Reply:
I think the fact that it returns the position is kind of an afterthought. I guess it ultimately just parses a word, so maybe evne just `parseWord`?

Reply:
Hmm. It does return the position though. How about `wordPos`?

Reply:
Right, but the main thing is the effect (i.e. the stuff that happens in `Parser`). If you call this function the main thing you want to know is that it will consume the given word from the input or fail... the fact that it returns the position is nice but less important.

Reply:
I see. But that's kinda what happens with parsers. I think adding more info in the name is helpful? In the same spirit as the parsers `lamTerm`, `appTerm` etc, `wordPos` tells us a bit about what the parser returns. It's probably not worth the time to think about it too much though! I'm happy to change it to either `parseWord` or `wordPos`! Lmk what you'll approve! :)

Reply:
and of course whatever @kwxm is happy with too.

Reply:
I think this is the only thing left? (Kenneth pointed out a couple other things, they are noted in the comments. I hope will be looked after in the next PR.)

Reply:
Just to continue the argument... ;) I think the type of `parseWord` should "really" be:
```
data Annotated ann a = Annotated ann a

parseWord :: Text -> Parser (Annotated SourcePos Text)
```
i.e. just like all the other parsers it returns the thing it parsed annotated with a position. It's just that in this case we know the thing it parsed statically, so we don't bother returning it.

Anyway: you pick a name, I don't mind too much.